Fix for #314

dpoetzsch opened the pull request:
This fixes issue #314.

The fix was to not check already validated properties against `additionalProperties` again. This assumes that `additionalProperties` is used exclusively for properties defined as `[key: type]: type`.

If `additionalProperties` was meant to set up additional restrictions on *all* properties please let me know, then the fix would be the allow `undefined` as value for type `any`. For my use case this would work as well, but it would break models like the following (which might be unintuitive to users):
```typescript
interface ExampleModel {
    a?: string;
    [key: string]: number;
}
```

endor commented on tests/unit/swagger/templateHelpers.spec.ts:
```diff
@@ -4,6 +4,62 @@ import * as templateHelpers from './../../../src/routeGeneration/templateHelpers
 
+    afterEach(() => {
+        // clean up global models
+        Object.keys(templateHelpers.models).forEach(k => {
```
why not simple reset it to `{}`?

dpoetzsch replied:
This is not possible in typescript because exported globals are read-only from outside the module.

I agree this is a bit hacky.

I think a cleaner approach would be to avoid the global and wrap all the validation functions into a service class that takes the model as constructor argument. But I did not want to change so much in your code.

endor commented on tests/unit/swagger/templateHelpers.spec.ts:
```diff
@@ -4,6 +4,62 @@ import * as templateHelpers from './../../../src/routeGeneration/templateHelpers
 
 describe('templateHelpers', () => {
 
+    afterEach(() => {
```
hm.. if we were to call `ValidateParam` in the unit tests instead of directly calling `validateModel`, wouldn't that save is this `afterEach` and also the exports above?

dpoetzsch replied:
`models` is accessed by `validateArray` as well, which is directly called from the tests. This is where such a dependency might occur.

I will just refactor this, removing the shared state, no problem.

endor replied:
Ah, you're right. I'm not sure those "sub"-validators should be exported. But thanks for fixing the shared state.

dpoetzsch replied:
Yeah, I'm always unsure as well. But I personally think in this case it makes sense to make them available so they can be tested individually because validation is quite complex for each case..

endor commented:
Can you explain what you mean when you say that `additionalProperties` could be meant to set up additional restrictions on all properties? The `ExampleModel` is illegal TypeScript, right?

dpoetzsch commented:
Actually your are right, it is illegal typescript, I did not see that, thanks.

What I meant is probably made more clear by the following test case:

```typescript
it('should validate all model properties against constraint defined in additional properties', () => {
    templateHelpers.models.ExampleModel = {
        additionalProperties: {
            dataType: 'integer',
            validators: { minimum: { value: 10 } }
        },
        properties: {
            a: { dataType: 'integer' },
        },
    };
    const error = {};
    templateHelpers.validateModel('', { a: 9 }, 'ExampleModel', error);
    expect(Object.keys(error)).not.to.be.empty;
});
```
Is this the expected outcome? If yes, my fix is not correct, yet.

endor commented:
Ah, good question. I think it's good to be explicit, so I'd say the above test case is not expected to work.

dpoetzsch commented:
I removed the global variable now and added the test above but with inverted expectation (expecting no error now). For backwards compatibility I also added a ValidateParam helper function that wraps instance creation.

 All tests work now.
